bin/propolis-utils/src/bin/cpuid-gen.rs

```diff
@@ -0,0 +1,249 @@
+            && self.edx == 0x69746e65
+    }
+    const fn all_zeros(&self) -> bool {
+        self.eax == 0 && self.ebx == 0 && self.ecx == 0 && self.edx == 0
+    }
+}
+impl From<&bhyve_api::vm_legacy_cpuid> for Cpuid {
+    fn from(value: &bhyve_api::vm_legacy_cpuid) -> Self {
+        Self {
+            eax: value.vlc_eax,
+            ebx: value.vlc_ebx,
+            ecx: value.vlc_ecx,
+            edx: value.vlc_edx,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Eq, PartialEq)]
+enum CpuidKey {
+    Leaf(u32),
+    SubLeaf(u32, u32),
+    LeafDefault(u32),
+}
+impl CpuidKey {
+    const fn eax(&self) -> u32 {
+        match self {
+            CpuidKey::Leaf(eax) => *eax,
+            CpuidKey::SubLeaf(eax, _) => *eax,
+            CpuidKey::LeafDefault(eax) => *eax,
+        }
+    }
+}
+impl Ord for CpuidKey {
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self.eq(other) {
+            return Ordering::Equal;
+        }
+        match self.eax().cmp(&other.eax()) {
+            Ordering::Equal => match (self, other) {
+                (CpuidKey::Leaf(_), _) | (_, CpuidKey::LeafDefault(_)) => {
+                    Ordering::Greater
+                }
+                (CpuidKey::LeafDefault(_), _) | (_, CpuidKey::Leaf(_)) => {
+                    Ordering::Less
+                }
+                (CpuidKey::SubLeaf(_, ecx), CpuidKey::SubLeaf(_, oecx)) => {
+                    ecx.cmp(oecx)
+                }
+            },
+            o => o,
+        }
+    }
+}
+impl PartialOrd for CpuidKey {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+fn query_cpuid(vm: &VmmFd, eax: u32, ecx: u32) -> anyhow::Result<Cpuid> {
+    let mut data = bhyve_api::vm_legacy_cpuid {
+        vlc_eax: eax,
+        vlc_ecx: ecx,
+        ..Default::default()
+    };
+    unsafe { vm.ioctl(bhyve_api::VM_LEGACY_CPUID, &mut data) }?;
+    Ok(Cpuid::from(&data))
+}
+
+const STD_EAX_BASE: u32 = 0x0;
+const EXTD_EAX_BASE: u32 = 0x80000000;
+
+const CPU_FEAT_ECX_XSAVE: u32 = 1 << 26;
+
+fn collect_cpuid(
+    vm: &VmmFd,
+    zero_elide: bool,
+) -> anyhow::Result<BTreeMap<CpuidKey, Cpuid>> {
+    let std = query_cpuid(vm, STD_EAX_BASE, 0)?;
+    let extd = query_cpuid(vm, EXTD_EAX_BASE, 0)?;
+
+    let mut results: BTreeMap<CpuidKey, Cpuid> = BTreeMap::new();
+
+    let mut xsave_supported = false;
+    let is_amd = std.is_authentic_amd();
+
+    for eax in 0..std.eax {
+        let data = query_cpuid(vm, eax, 0)?;
+
+        if zero_elide && data.all_zeros() {
+            continue;
+        }
+
+        match eax {
+            0x1 => {
+                if data.ecx & CPU_FEAT_ECX_XSAVE != 0 {
+                    xsave_supported = true;
+                }
+                results.insert(CpuidKey::Leaf(eax), data);
+            }
+            0x7 => {
+                results.insert(CpuidKey::SubLeaf(eax, 0), data);
+
+                // TODO: handle more sub-leaf entries?
```

Comment:
Worth using the `x86` crate so we don't have to manually handle every case here? It does seem to support just providing a custom function instead of trying to execute cpuid itself: https://docs.rs/x86/latest/x86/cpuid/struct.CpuId.html#method.with_cpuid_fn

Reply:
While some of the handling in that crate looks convenient, I'm not sure it's what we want here, since it seems less focused on manipulating the raw values?